### PR TITLE
Create better messages for auto-indexing

### LINKED ISSUE #309: Indexed and positional CodeBlock replacement blocks error
..is awful. I don't have an example right now, but both too few and too many arguments create misleading exceptions based on list iteration.

### PR DESCRIPTION
**Pull request by kenzierocks**
Fixes #309.

I'm not really sure about checking for unused arguments, it seems a little useless. If it is preferred, I'll write a test for it.

### CODE REVIEW

**Comment by JakeWharton** [comment]
Unused arguments are an error.

**Comment by JakeWharton** [comment]
Also mixing indexed and positional should be flat out forbidden as it gets crazy fast.

**Comment by kenzierocks** [comment]
Ok, I'll whip up a test then.

**Comment by kenzierocks** [comment]
Ok, done.

**Comment by JakeWharton** [comment]
Wow you're fast! Thanks.

File: src/test/java/com/squareup/javapoet/TypeSpecTest.java [hunk 1]
```diff
@@ -1848,7 +1848,17 @@ private String toString(TypeSpec typeSpec) {
...
+      assertThat(expected).hasMessage("Not enough parameters were provided; expected at least 1, got 0");
+    }
+  }
+
+  @Test public void unusedArguments() {
+    try {
+      CodeBlock.builder().add("", "foo");
+      fail();
+    } catch (IllegalArgumentException expected) {
+      assertThat(expected).hasMessage("Too many parameters were given; parameters [foo] were unused");
```
Comment (JakeWharton) [inline src/test/java/com/squareup/javapoet/TypeSpecTest.java]: I think this message is a bit much. Can we make them match?
> "Expected 1, got 0"
> "Expected 0, got 1"

**Comment by kenzierocks** [comment]
@JakeWharton I'll remove support for that then, and probably change how the too many arguments detector works.

File: src/test/java/com/squareup/javapoet/TypeSpecTest.java (hunk 1 above)
Comment (kenzierocks) [inline src/test/java/com/squareup/javapoet/TypeSpecTest.java]: Yes, I'll do that with the removal of index with positional.

**Comment by kenzierocks** [comment]
Alright, that should be good. Tell me if there's anything you want changed :)

File: src/test/java/com/squareup/javapoet/TypeSpecTest.java [hunk 2]
```diff
@@ -1848,7 +1848,17 @@ private String toString(TypeSpec typeSpec) {
     try {
       CodeBlock.builder().add("$S");
       fail();
-    } catch (NoSuchElementException expected) {
+    } catch (IllegalArgumentException expected) {
+      assertThat(expected).hasMessage("Not enough parameters were given; expected 1, got 0");
+    }
+  }
+
+  @Test public void unusedArguments() {
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/TypeSpecTest.java]: Nice.

**Comment by swankjesse** [comment]
Neato.

**Pull request merged**